src/backend/auxil/range_alloc.rs

```diff
@@ -0,0 +1,190 @@
+use std::fmt::Debug;
+use std::ops::{Add, AddAssign, Range, Sub};
+
+pub struct RangeAllocator<T> {
+    /// The range this allocator covers.
+    initial_range: Range<T>,
+    /// A Vec of ranges in this heap which are unused.
+    /// Must be ordered with ascending range start to permit short circuiting allocation.
+    /// No two ranges in this vec may overlap.
+    free_ranges: Vec<Range<T>>,
+}
+
+impl<T> RangeAllocator<T>
+where
+    T: Clone + Copy + Add<Output = T> + AddAssign + Sub<Output = T> + Eq + PartialOrd + Debug,
+{
+    pub fn new(range: Range<T>) -> Self {
+        RangeAllocator {
+            initial_range: range.clone(),
+            free_ranges: vec![range.clone()],
+        }
+    }
+
+    pub fn initial_range(&self) -> Range<T> {
+        self.initial_range.clone()
+    }
+
+    pub fn allocate_range(&mut self, length: T) -> Option<Range<T>> {
+        let mut best_fit: Option<(usize, Range<T>)> = None;
+        for range in self.free_ranges.iter().cloned().enumerate() {
+            let (index, range) = range;
+            let range_length = range.end - range.start;
+            if range_length < length {
+                continue;
+            } else if range_length == length {
+                // Found a perfect fit, so stop looking.
+                best_fit = Some((index, range));
+                break;
+            }
+            match best_fit {
+                Some((ref mut best_index, ref mut best_range)) => {
+                    // Find best fit for this allocation to reduce memory fragmentation.
+                    if range_length < best_range.end - best_range.start {
+                        *best_index = index;
+                        *best_range = range.clone();
+                    }
+                }
+                None => {
+                    best_fit = Some((index, range.clone()));
+                }
+            }
+        }
+        best_fit.map(|range| {
+            let (index, range) = range;
+            if range.end - range.start == length {
```

Comment:
`range.end - range.start` can be replaced by `range_length`

Reply:
It can't, because `range_length` is out of scope here.